src/System.Management.Automation/engine/MshMemberInfo.cs

```diff
@@ -4309,10 +4319,10 @@ public override IEnumerator<T> GetEnumerator()
                 return Enumerable.Empty<T>().GetEnumerator();
             }
 
-            lock (Members)
+            lock (_members)
```

Comment:
@daxian-dbw Please clarify - is it ok to lock on different objects in different methods? What if different threads call Add() and Remove() in the same time?

Reply:
All locks are essentially on the same object. When `Add` and `Remove` are called at the same time, there are two possibilities:
1. The check of `_members == null` in `Remove` happens **before** the initialization of `_member` triggered by `Add`;
1. The check of `_members == null` in `Remove` happens **after** the initialization of `_member` triggered by `Add`.

In the first case, `Remove` will just return, and `Add` will lock on the newly initialized object.
In the second case, `Remove` will try to lock on `_members` (once it's not null, it will never be set back to null), which points the same object as `Member`, so `Remove` and `Add` will compete locking on the same object.